### ISSUE TITLE
[2015.8.0rc2] pkg.installed performance regression

### ISSUE THREAD

**Issue by tbaker57** [saltstack/salt#25776]
I'm not sure what's going on here. With a simple state
```
httpd_package:
  pkg:
    - installed
    - name: httpd

# we always have mod_ssl
mod_ssl:
  pkg:
    - installed
    - require:
      - pkg: httpd_package
```
running this state on a fresh Centos 6.6 vagrant box with 2015.5.2 yields
```
time salt-call --local state.sls httpd.httpd-deploy -ldebug
...
[DEBUG   ] Results of YAML rendering: 
OrderedDict([('httpd_package', OrderedDict([('pkg', ['installed', OrderedDict([('name', 'httpd')])])])), ('mod_ssl', OrderedDict([('pkg', ['installed', OrderedDict([('require', [OrderedDict([('pkg', 'httpd_package')])])])])]))])
[DEBUG   ] LazyLoaded pkg.install
[DEBUG   ] LazyLoaded pkg.mod_init
[DEBUG   ] Error loading module.npm: npm execution module could not be loaded because the npm binary could not be located
[DEBUG   ] Error loading module.tls: ['PyOpenSSL must be installed before this module can be used.']
[DEBUG   ] Error loading module.nacl: libnacl import error, perhaps missing python libnacl package
[DEBUG   ] Error loading module.ipmi: No module named pyghmi.ipmi
[DEBUG   ] Could not LazyLoad pkg.ex_mod_init
[INFO    ] Running state [httpd] at time 01:50:34.610880
[INFO    ] Executing state pkg.installed for httpd
[INFO    ] Executing command "repoquery --plugins --queryformat '%{NAME}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-%{REPOID}' --all --pkgnarrow=installed" in directory '/root'
[INFO    ] Executing command "repoquery --plugins --queryformat '%{NAME}_|-%{ARCH}'  --pkgnarrow=all --all" in directory '/root'
[INFO    ] Executing command "repoquery --plugins --queryformat '%{NAME}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-%{REPOID}'   --all --quiet --whatprovides httpd" in directory '/root'
[INFO    ] Executing command ['yum', '-q', 'clean', 'expire-cache'] in directory '/root'
[DEBUG   ] output: 
[INFO    ] Executing command ['yum', '-q', 'check-update'] in directory '/root'
[DEBUG   ] output: 
dracut.noarch                         004-356.el6_6.3                    updates
dracut-kernel.noarch                  004-356.el6_6.3                    updates
kernel.x86_64                         2.6.32-504.30.3.el6                updates
kernel-devel.x86_64                   2.6.32-504.30.3.el6                updates
kernel-firmware.noarch                2.6.32-504.30.3.el6                updates
kernel-headers.x86_64                 2.6.32-504.30.3.el6                updates
[INFO    ] Executing command 'yum -y     install httpd' in directory '/root'
[INFO    ] Executing command "repoquery --plugins --queryformat '%{NAME}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-%{REPOID}' --all --pkgnarrow=installed" in directory '/root'
[INFO    ] Installed Packages:
httpd changed from absent to 2.2.15-39.el6.centos

[DEBUG   ] Refreshing modules...
[INFO    ] Loading fresh modules for state activity
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[INFO    ] Completed state [httpd] at time 01:50:40.370691
[DEBUG   ] LazyLoaded config.option
[INFO    ] Running state [mod_ssl] at time 01:50:40.371535
[DEBUG   ] LazyLoaded pkg.install
[DEBUG   ] LazyLoaded pkg.installed
[INFO    ] Executing state pkg.installed for mod_ssl
[DEBUG   ] LazyLoaded cmd.run_all
[INFO    ] Executing command "repoquery --plugins --queryformat '%{NAME}_|-%{ARCH}'  --pkgnarrow=all --all" in directory '/root'
[INFO    ] Executing command "repoquery --plugins --queryformat '%{NAME}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-%{REPOID}'   --all --quiet --whatprovides mod_ssl" in directory '/root'
[DEBUG   ] LazyLoaded pkg_resource.parse_targets
[INFO    ] Executing command 'yum -y     install mod_ssl' in directory '/root'
[INFO    ] Executing command "repoquery --plugins --queryformat '%{NAME}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-%{REPOID}' --all --pkgnarrow=installed" in directory '/root'
[INFO    ] Installed Packages:
mod_ssl changed from absent to 2.2.15-39.el6.centos

[DEBUG   ] Refreshing modules...
[INFO    ] Loading fresh modules for state activity
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[INFO    ] Completed state [mod_ssl] at time 01:50:43.067828
[DEBUG   ] File /var/cache/salt/minion/accumulator/31692816 does not exist, no need to cleanup.
[DEBUG   ] LazyLoaded highstate.output
[DEBUG   ] LazyLoaded nested.output
[DEBUG   ] LazyLoaded nested.output
local:
----------
          ID: httpd_package
    Function: pkg.installed
        Name: httpd
      Result: True
     Comment: The following packages were installed/updated: httpd
     Started: 01:50:34.610880
    Duration: 5759.811 ms
     Changes:   
              ----------
              httpd:
                  ----------
                  new:
                      2.2.15-39.el6.centos
                  old:
----------
          ID: mod_ssl
    Function: pkg.installed
      Result: True
     Comment: The following packages were installed/updated: mod_ssl
     Started: 01:50:40.371535
    Duration: 2696.293 ms
     Changes:   
              ----------
              mod_ssl:
                  ----------
                  new:
                      2.2.15-39.el6.centos
                  old:

Summary
------------
Succeeded: 2 (changed=2)
Failed:    0
------------
Total states run:     2

real	0m10.341s
user	0m7.245s
sys	0m1.648s

```
whereas running it on 2015.8.0rc2 yields:
```
[DEBUG   ] Results of YAML rendering: 
OrderedDict([('httpd_package', OrderedDict([('pkg', ['installed', OrderedDict([('name', 'httpd')])])])), ('mod_ssl', OrderedDict([('pkg', ['installed', OrderedDict([('require', [OrderedDict([('pkg', 'httpd_package')])])])])]))])
[DEBUG   ] LazyLoaded pkg.install
[DEBUG   ] LazyLoaded pkg.installed
[DEBUG   ] Error loading module.victorops: No VictorOps api key found.
[DEBUG   ] Error loading module.ifttt: IFTTT Secret Key Unavailable, not loading.
[DEBUG   ] Error loading module.tls: ['PyOpenSSL version 0.10 or later must be installed before this module can be used.']
[DEBUG   ] Error loading module.nacl: libnacl import error, perhaps missing python libnacl package
[DEBUG   ] Error loading module.ipmi: No module named pyghmi.ipmi
[DEBUG   ] Error loading module.npm: npm execution module could not be loaded because the npm binary could not be located
[DEBUG   ] Could not LazyLoad pkg.ex_mod_init
[INFO    ] Running state [httpd] at time 02:02:13.152223
[INFO    ] Executing state pkg.installed for httpd
[INFO    ] Executing command "repoquery --plugins --queryformat '%{NAME}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-%{REPOID}' --all --pkgnarrow=installed" in directory '/root'
[INFO    ] Executing command "repoquery --plugins --queryformat '%{NAME}_|-%{ARCH}'  --pkgnarrow=all --all" in directory '/root'
[INFO    ] Executing command "repoquery --plugins --queryformat '%{NAME}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-%{REPOID}'   --all --quiet --whatprovides httpd" in directory '/root'
[INFO    ] Executing command ['yum', '-q', 'clean', 'expire-cache'] in directory '/root'
[DEBUG   ] output: 
[INFO    ] Executing command ['yum', '-q', 'check-update'] in directory '/root'
[DEBUG   ] output: 
dracut.noarch                         004-356.el6_6.3                    updates
dracut-kernel.noarch                  004-356.el6_6.3                    updates
kernel.x86_64                         2.6.32-504.30.3.el6                updates
kernel-devel.x86_64                   2.6.32-504.30.3.el6                updates
kernel-firmware.noarch                2.6.32-504.30.3.el6                updates
kernel-headers.x86_64                 2.6.32-504.30.3.el6                updates
[INFO    ] Executing command 'yum -y     install httpd' in directory '/root'
[INFO    ] Executing command "repoquery --plugins --queryformat '%{NAME}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-%{REPOID}' --all --pkgnarrow=installed" in directory '/root'
[INFO    ] Made the following changes:
httpd changed from absent to 2.2.15-39.el6.centos

[DEBUG   ] Refreshing modules...
[INFO    ] Loading fresh modules for state activity
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[INFO    ] Completed state [httpd] at time 02:02:44.937789
[DEBUG   ] LazyLoaded config.option
[DEBUG   ] LazyLoaded pkg.install
[DEBUG   ] LazyLoaded pkg.installed
[INFO    ] Running state [mod_ssl] at time 02:02:44.940192
[INFO    ] Executing state pkg.installed for mod_ssl
[DEBUG   ] LazyLoaded cmd.run_all
[INFO    ] Executing command "repoquery --plugins --queryformat '%{NAME}_|-%{ARCH}'  --pkgnarrow=all --all" in directory '/root'
[INFO    ] Executing command "repoquery --plugins --queryformat '%{NAME}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-%{REPOID}'   --all --quiet --whatprovides mod_ssl" in directory '/root'
[DEBUG   ] LazyLoaded pkg_resource.parse_targets
[INFO    ] Executing command 'yum -y     install mod_ssl' in directory '/root'
[INFO    ] Executing command "repoquery --plugins --queryformat '%{NAME}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-%{REPOID}' --all --pkgnarrow=installed" in directory '/root'
[INFO    ] Made the following changes:
mod_ssl changed from absent to 2.2.15-39.el6.centos

[DEBUG   ] Refreshing modules...
[INFO    ] Loading fresh modules for state activity
[DEBUG   ] LazyLoaded jinja.render
[DEBUG   ] LazyLoaded yaml.render
[INFO    ] Completed state [mod_ssl] at time 02:03:01.210813
[DEBUG   ] File /var/cache/salt/minion/accumulator/43805968 does not exist, no need to cleanup.
[DEBUG   ] LazyLoaded highstate.output
[DEBUG   ] LazyLoaded nested.output
[DEBUG   ] LazyLoaded nested.output
local:
----------
          ID: httpd_package
    Function: pkg.installed
        Name: httpd
      Result: True
     Comment: The following packages were installed/updated: httpd
     Started: 02:02:13.152223
    Duration: 31785.566 ms
     Changes:   
              ----------
              httpd:
                  ----------
                  new:
                      2.2.15-39.el6.centos
                  old:
----------
          ID: mod_ssl
    Function: pkg.installed
      Result: True
     Comment: The following packages were installed/updated: mod_ssl
     Started: 02:02:44.940192
    Duration: 16270.621 ms
     Changes:   
              ----------
              mod_ssl:
                  ----------
                  new:
                      2.2.15-39.el6.centos
                  old:

Summary for local
------------
Succeeded: 2 (changed=2)
Failed:    0
------------
Total states run:     2
Total run time:   48.056 s

real	0m49.990s
user	0m40.409s
sys	0m6.425s
```
which is a significant performance difference. The yum calls do go via a forward proxy, but I've confirmed that all rpms are served from the cache, so network issues don't appear to be involved.

There is not a lot of timing information in the debug output. However, visually, it appears that the 2 calls 
```
[INFO    ] Executing command "repoquery --plugins --queryformat '%{NAME}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-%{REPOID}' --all --pkgnarrow=installed" in directory '/root'
```
are the ones taking a lot longer in the 2015.8.0rc2 case.

cheers,
Tim

**Comment by jfindlay** [comment]
@tbaker57, thanks for your excellent work on this.

**Comment by basepi** [comment]
@terminalmage Could this be related to your recent additions to the pkg modules? Can't remember the exact details of your additions, but I know you've been in there recently.

**Comment by terminalmage** [comment]
I doubt it, there shouldn't be anything that would increase the amount of time it takes to run to that extent.

@tbaker57 If you uninstall the packages without clearing the yum cache, and the states are re-run, does it still take as long? If not, then this could point to mirror issues.

**Comment by terminalmage** [comment]
Yeah, this really looks like a yum caching/mirror difference. I'm getting the same longer length for the initial run on the 2015.5 branch, but after the first run I am getting speeds similar to the 2015.5.2 run @tbaker57 posted.

I don't think this is a Salt issue.

**Comment by tbaker57** [comment]
@terminalmage - thanks for trying to reproduce. Yes, I'm just doing a 'yum erase httpd mod_ssl' (which I don't think clears the cache) and the time remains the same.

I've just run this under strace and at the point where it bottlenecks, I can see heaps of attempts to open up the (non-existent) file 'x86_64'
```
--- SIGCHLD (Child exited) @ 0 (0) ---
pipe([4, 5])                            = 0
pipe([6, 7])                            = 0
pipe([8, 9])                            = 0
fcntl(9, F_GETFD)                       = 0
fcntl(9, F_SETFD, FD_CLOEXEC)           = 0
clone(child_stack=0, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f63bb61d9d0) = 14374
close(9)                                = 0
close(5)                                = 0
close(7)                                = 0
read(8, "", 1048576)                    = 0
close(8)                                = 0
fcntl(4, F_GETFL)                       = 0 (flags O_RDONLY)
fstat(4, {st_mode=S_IFIFO|0600, st_size=0, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f63bb627000
lseek(4, 0, SEEK_CUR)                   = -1 ESPIPE (Illegal seek)
fstat(4, {st_mode=S_IFIFO|0600, st_size=0, ...}) = 0
munmap(0x7f63bb627000, 4096)            = 0
fcntl(6, F_GETFL)                       = 0 (flags O_RDONLY)
fstat(6, {st_mode=S_IFIFO|0600, st_size=0, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f63bb627000
lseek(6, 0, SEEK_CUR)                   = -1 ESPIPE (Illegal seek)
fstat(6, {st_mode=S_IFIFO|0600, st_size=0, ...}) = 0
munmap(0x7f63bb627000, 4096)            = 0
poll([{fd=4, events=POLLIN|POLLPRI}, {fd=6, events=POLLIN|POLLPRI}], 2, -1) = 1 ([{fd=4, revents=POLLIN}])
read(4, "x86_64\n", 4096)               = 7
poll([{fd=4, events=POLLIN|POLLPRI}, {fd=6, events=POLLIN|POLLPRI}], 2, -1) = 1 ([{fd=4, revents=POLLHUP}])
close(4)                                = 0
poll([{fd=6, events=POLLIN|POLLPRI}], 1, -1) = 1 ([{fd=6, revents=POLLHUP}])
close(6)                                = 0
wait4(14374, [{WIFEXITED(s) && WEXITSTATUS(s) == 0}], 0, NULL) = 14374
--- SIGCHLD (Child exited) @ 0 (0) ---
```
which seems to indicate some parsing issue somewhere. Anyway, I'll see if I can dig a little further this weekend. If I can't get anywhere, I'm happy to close this...

cheers,
Tim

**Comment by tbaker57** [comment]
Ok - I stole some company time on a quiet Friday afternoon and figured it out with the help of lots of print statements.

The problem lies here:
https://github.com/saltstack/salt/blob/v2015.8.0rc2/salt/utils/pkg/rpm.py#L81
```
def parse_pkginfo(line, osarch=None):
...
    if not check_32(arch):
        if arch not in (osarch, 'noarch'):
            name += '.{0}'.format(arch)
```
parse_pkginfo is called for every package installed on the system (469 in my case). The initial os_arch='x86_64' is passed to this call. The definition of check_32 is
```
def check_32(arch, osarch=None):
    '''
    Returns True if both the OS arch and the passed arch are 32-bit
    '''
    if osarch is None:
        osarch = _osarch()
    return all(x in ARCHES_32 for x in (osarch, arch))
```
and if osarch is None _osarch() goes and makes an expensive subprocess.Popen call to 'rpm --eval'.

The issue is that L81 is calling check_32(arch), not check_32(arch, osarch), so for all 469 packages an expensive subprocess.Popen call is made to 'rpm --eval ....'
Changing
```
81c81
<     if not check_32(arch, osarch):
---
>     if not check_32(arch):
```
gives me similar times for 2015.5.2 and 2015.8.0rc2

cheers,
T2

**Comment by terminalmage** [comment]
Well, that would certainly explain why I was unable to reproduce, my tests were on a minimal install. Thanks very much for your excellent investigative work!

I submitted your fix in #25910.

Thanks again!

**Comment by tbaker57** [comment]
Happy to help. We gain a lot from using Salt here, so it's only fair to contribute something back in testing.

cheers,
Tim
